### ISSUE TITLE
Flatten macro expansions

### ISSUE THREAD

**Issue by fpoli** [fpoli/viper-silver#137]
Whenever a macro is expanded in the body of a method, a new Seqn node \(AST\) is generally created as a placeholder for all statements from the macro. As a result the body of the method is a Seqn containing another Seqn. Worse, in cases of multiply nested macro expansions as in nestedBlocks.sil \(in attachment\), we can have multiple layers of Seqn composed of Seqn.

A side effect of such nesting can be seen in the transformation framework, particularly when neighbor nodes are used as a context for detecting when to rewrite a statement. For an example, refer to ManyToOneAssert test case in RewriterTest.scala.

Currently, the AST is left as is after nested macro expansions and it is the job of the rewriter framework to tackle nested statements and linearize them prior to rewriting.

This should be reverted: the rewriter framework should be agnostic to macro expansions that might have taken place previously. And the macro expansion should deliver a linearized sequence of statements \(Seqn\) without unnecessary nesting.

[Issue created by fabiopakk: 2019-07-02]
[Last updated on bitbucket: 2019-07-10]

[Comment created by mschwerhoff: 2019-07-02]
Quick note: Since front-ends can create arbitrary ASTs, the transformation framework must still be able to handle nested `Seqn` nodes, regardless of how macro expansion is implemented.

[Comment created by fpoli_eth: 2019-07-02]
I confirm that in Prusti we generate nested `Seqn` nodes; it's quite convenient for us.

[Comment created by mschwerhoff: 2019-07-02]
Voila generates nested `Seqn` as well.

[Comment created by fabiopakk: 2019-07-03]
It is not a problem to generate nested Seqn, as long as they are equivalent to their flattened versions. 

The issue is that the transformation framework has a cumbersome code just to make the context’s successors and predecessors work, when in presence of nested Seqn. One has to decide if successors, for example, mean a nested Seqn or the statements in its flattened version.

However, the scope of this task is macro expansion. But it can easily be adopted for every situation. Would you see any consequences in flattening all nested Seqn prior to any processing? This could be done in another task.

[Comment created by mschwerhoff: 2019-07-04]
Flattening `Seqn` by default might surprise users. Using nested `Seqn` can be convenient for front-end developers, e.g. to structure the AST they create. If they create an AST  `val ast = Seqn(Seqn(...), Seqn(...))`, then replace all variables `x` instead `ast` with some `y` and then pattern-match on the AST via `val Seqn(left, right) = ast`, they’ll most likely won’t expect a pattern match failure.

In my opinion, flattening is an AST optimisation, and as such should be explicitly performed by users, not automatically for technical reasons.

[Comment created by fabiopakk: 2019-07-08]
I see. I’ll deploy the solution to this issue \(limited to the scope of macro expansions\). And then lets check if there’s any interference with the frontends. The effects should be perpendicular.

[Comment created by fabiopakk: 2019-07-10]
Indeed, this example is what is described in the description of this issue, only the other way around. So this discussion is very good, since the issue seems to make sense particularly after macro expansions but not generally. So I’ll make sure it is restricted.

[Comment created by fabiopakk: 2019-07-10]
Issue fixed in branch fabiopakk_issue_279_flatten_macro_expansions, already merged into default branch.

**Comment by fpoli** [comment]
> **@mschwerhoff** commented on 2016-04-12 12:17

```
#!text

Exception in thread "main" java.lang.ClassCastException: viper.silver.parser.PMethod cannot be cast to viper.silver.parser.PAnyFunction
	at viper.silver.parser.TypeChecker.check(Resolver.scala:513)
	at viper.silver.parser.TypeChecker.checkTopTyped(Resolver.scala:432)
	at viper.silver.parser.TypeChecker.check(Resolver.scala:429)
	at viper.silver.parser.TypeChecker.check(Resolver.scala:171)
	at viper.silver.parser.TypeChecker$$anonfun$check$14.apply(Resolver.scala:155)
	at viper.silver.parser.TypeChecker$$anonfun$check$14.apply(Resolver.scala:155)
        ...
```

Probably related to #127

**Comment by fpoli** [comment]
> **@meilers** commented on 2016-07-29 12:47

Resolved in pull request #54.
